src/visualizers/VisualizerChaos.ts

```diff
@@ -0,0 +1,393 @@
+		}
+		this.backgroundColor = sketch.color(myHexBack);
+		this.textColor = sketch.color(myHexText);
+	}
+}
+
+enum ColorStyle { Walker, Corner, Index, Highlight }
+
+// params schema
+const schemaChaos = [
+	new VisualizerParamsSchema(
+		"num",
+		ParamType.number,
+		"Number of terms",
+		true,
+		10000,
+		"The number of terms to stop drawing at (a value of 0 will draw forever/until it runs out)."
+	),
+	new VisualizerParamsSchema(
+		"corners",
+		ParamType.number,
+		"Number of corners",
+		true,
+		4,
+		"The number of corners on the polygon, also the modulus we apply to the terms of the sequence."
+	),
+	new VisualizerParamsSchema(
+		"walkers",
+		ParamType.number,
+		"Number of walkers",
+		true,
+		1,
+		"The number w of walkers; break into subsequences based on residue mod w of the index, each with a separate walker."
+	),
+	// it would be nice to change this to an enum, if the params schema allowed
+	new VisualizerParamsSchema(
+		"style",
+		ParamType.number,
+		"Color scheme",
+		true,
+		1,
+		"0 = colour by walker; 1 = colour by destination; 2 = colour by index; 3 = colour one walker."
+	),
+	new VisualizerParamsSchema(
+		"highlightWalker",
+		ParamType.number,
+		"Highlighted walker",
+		false,
+		0,
+		"The walker to highlight in colour scheme 3."
+	),
+	new VisualizerParamsSchema(
+		"circSize",
+		ParamType.number,
+		"Circle size",
+		true,
+		1,
+		"Size of the dots."
+	),
+	new VisualizerParamsSchema(
+		"alpha",
+		ParamType.number,
+		"Circle alpha",
+		true,
+		1,
+		"Alpha factor (transparency, 0=transparent, 1=solid) of the dots."
+	),
+	new VisualizerParamsSchema(
+		"frac",
+		ParamType.number,
+		"Fraction to walk",
+		true,
+		0.5,
+		"How far each step takes you toward the corner (value between 0 and 1 inclusive)."
+	),
+	new VisualizerParamsSchema(
+		"pixelsPerFrame",
+		ParamType.number,
+		"Dots per frame",
+		true,
+		400,
+		"How many dots to draw per frame (more = faster)."
+	),
+	new VisualizerParamsSchema(
+		"showLabels",
+		ParamType.boolean,
+		"Show corner labels?",
+		true,
+		false,
+		"Whether to label corners of polygon."
+	),
+	new VisualizerParamsSchema(
+		"darkMode",
+		ParamType.boolean,
+		"Use dark mode?",
+		true,
+		false,
+		"Whether to make a dark background."
+	)
+	];
+	// other ideas:  previous parts of the sequence fade over time, or shrink over time
+	// circles fade to the outside
+
+
+class VisualizerChaos extends VisualizerDefault {
+    
+	name = "Chaos";
+
+	private img: p5.Image = new p5.Image();
+
+	// private properly typed versions of the user parameters
+	private num = 0;
+	private corners = 0;
+	private walkers = 0;
+	private style = 0;
+	private colorStyle = ColorStyle.Walker;
+	private highlightWalker = 0;
+	private circSize = 0;
+	private alpha = 1;
+	private frac = 0;
+	private pixelsPerFrame = 0;
+	private showLabels = false;
+	private darkMode = false;
+	
+	// current state variables (used in setup and draw)
+	private myIndex = 0;
+	private pixelCount = 0;
```

Comment:
Funny, yes.  This made me realize, though, that in future versions they may want to diverge because we may have an offset in the sequence; semantically they are different even if they never diverged in my code.  But then I realized I should handle that directly in a way that will be easy to update when we have offsets in our sequences.  Hence I replaced "number of terms" in the parameters with "first index" and "last index".  I removed `pixelCount` and dealt with things in terms of those parameters instead.  When we add the capability for offsets, we'll have to read the valid range of indices and then in validate(), we'll need to compare against the parameters and act accordingly.

Reply:
OK, so the infrastructure to have "offsets" in the sequences is already there in the main branch, which you'll need to rebase on to merge this visualizer (I think you already did the rebase): a Sequence object now has .first and .last properties, and you should only use getElement with indices between these values (inclusive). (Moreover as soon as the PR #80 is merged, then OEIS sequences will actually start reporting their .first and .last indices properly.)

So I think the only new thing you need to do is have this.offset default to this.seq.first rather than 0; and you could also contemplate whether (a) you want to bother allowing the user to specify an initial index, or just always go from this.seq.first; and even (b) you want to bother allowing the user to specify a final index, or just always keep running until this.seq.last, figuring that the user could just have constructed the sequence to end where desired.

Reply:
Ok missed this, I'll do it now.

Reply:
Ok, done, now it ought to only request terms between first and last.

Reply:
Hmmm, I don't think so -- if num is left at 0, and say this.seq.last is 10, then the min of 0 and 10 is 0, and then as far as I can tell the code just keeps on going after the item at index 10.  Please check that out and fix if so. (Frankly, I don't think you will be able to manage to use 0 as a value with special meaning in this way, given that there are sequences that start at negative indices; you will want to be able to iterate over them from the beginning, going through 0 and stopping at some positive index.)

Also, while you're at it, to my eye the variable names "offset" and "num" don't cry out "the first index I will visit" and "the last index I will visit". If you agree, then please change those variable names to something more easily understood.

When you're done with these items, I'll do another (hopefully final) full review. Thanks!

Reply:
Another thought: you could use `Infinity` as the value for the last index you will visit if the user left it blank (if the user put in a zero, you probably want to respect that given the possibility of negative indices), and then that will "min" with this.seq.last properly. Of course there are multiple ways to row this boat, so if you come up with another you like that has the correct behavior, that'll be fine.

Reply:
I didn't realize they could be negative!  But of course, we should handle that.  Ok, I've redesigned, and it made the logic simpler.  In the process, my attention was drawn to a few other things and I've reordered the user parameters, added a "colour cycle length" for when we are colouring by index but it is running forever.  Note when testing that it is running forever, that the thumbnail card doesn't, but the main view will.

Reply:
OK, looks like the limits are correctly respected now.